runtime/parachains/src/shared.rs

```diff
@@ -90,6 +168,9 @@ impl<T: Config> Pallet<T> {
 		new_config: &HostConfiguration<T::BlockNumber>,
 		all_validators: Vec<ValidatorId>,
 	) -> Vec<ValidatorId> {
+		// Drop allowed relay parents buffer on a session change.
+		AllowedRelayParents::<T>::mutate(|tracker| tracker.buffer.clear());
```

Comment:
This seems correct. The behavior is somewhat nuanced.

`on_new_session` is invoked in the `on_finalize` logic of the block that changes the session. So when we call `on_initialize` in the next block, it'll add a relay-parent to the tracker which is the last block of the prior session. This is fine, because we say that a candidate with a relay-parent R should be backed with the group assigned in number(R) + 1, which is guaranteed to be in the current session.

Sounds right?

Reply:
> 

Yes,
To put it simple, we always allow building on top of the previous block
With async backing it will be some lookback which is kind of disabled on session boundaries

Probably worth leaving a note about it in comments?

Reply:
Yeah, comments are always helpful.